### ISSUE TITLE
Error due to reading of structs after call to free

### ISSUE THREAD

**Issue by gliwka** [cerebuild/hyperscan-java#31]
Bug report by @eliaslevy in #28:

> A bit more on the error I was seeing.  I've reproduced it on my MBP and narrowed it down to a call I make to `Expression.validate` on start up to check the expressions.  It only seems to trigger if there are sufficient concurrent threads, each of which run the same code.   In my case I need at least four threads for it to trigger, and it won't trigger if I run the code in the debugger.  The error goes away if I synchronize the call to `Expression.validate`.
> 
> The error report generated by the JVM shows two threads calling into `hs_free_compile_error` from `validate` with the top of the stack being:
> 
> ```
> Stack: [0x00007000055c4000,0x00007000056c4000],  sp=0x00007000056c2440,  free space=1017k
> Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
> C  [libsystem_c.dylib+0x1b52]  strlen+0x12
> C  [jna4845073464406839331.tmp+0x4073]  Java_com_sun_jna_Native_getStringBytes+0x23
> j  com.sun.jna.Native.getStringBytes(Lcom/sun/jna/Pointer;JJ)[B+0
> j  com.sun.jna.Native.getString(Lcom/sun/jna/Pointer;JLjava/lang/String;)Ljava/lang/String;+6
> j  com.sun.jna.Pointer.getString(JLjava/lang/String;)Ljava/lang/String;+3
> j  com.sun.jna.Structure.readField(Lcom/sun/jna/Structure$StructField;)Ljava/lang/Object;+134
> j  com.sun.jna.Structure.read()V+100
> j  com.sun.jna.Structure.autoRead()V+8
> j  com.sun.jna.Function.invoke(Ljava/lang/reflect/Method;[Ljava/lang/Class;Ljava/lang/Class;[Ljava/lang/Object;Ljava/util/Map;)Ljava/lang/Object;+395
> j  com.sun.jna.Library$Handler.invoke(Ljava/lang/Object;Ljava/lang/reflect/Method;[Ljava/lang/Object;)Ljava/lang/Object;+344
> j  com.sun.proxy.$Proxy3.hs_free_compile_error(Lcom/gliwka/hyperscan/jna/CompileErrorStruct;)I+24
> j  com.gliwka.hyperscan.wrapper.Expression.validate()Lcom/gliwka/hyperscan/wrapper/Expression$ValidationResult;+77
> ```
> 
> Looks like it is dying in the call to `strlen` from JNA's `Java_com_sun_jna_Native_getStringBytes`.
> 
> I am not familiar with JNA, but I believe the issue is that `CompileErrorStruct` is created by passing it a native struct that has been allocated by hyperscan.  The call to `hs_free_compile_error` deallocates the the struct.  But JNA's `Structure` has a feature enabled by default that will read the contents of the native struct into the Java `Structure` after every native call.  So after calling `hs_free_compile_error`, JNA is trying to read the fields of the struct that hyperscan just freed.  
> 
> In the single threaded case this is usually not a problem, as nothing will have modified that memory.  But in a multi-threaded environment, the memory could be allocated and modified by another thread, leading to an error when the original thread tried to read the fields.  Something changed between hyperscan 4.5.2 and 4.6.0 to make this more likely, but the problem was present before we upgraded to 4.6.0.
> 
> If I modify hyperscan-java to call `errorStruct.setAutoRead(false);` before every call to `hs_free_compile_error`, thus disabling read of the struct after the call to `hs_free_compile_error`, the error goes away.
> 
> 
>

**Comment by eliaslevy** [comment]
Here is a [thread](https://groups.google.com/forum/m/#!topic/jna-users/OzB5yOy1-fU) that discusses this issue.

**Comment by eliaslevy** [comment]
Not that the problem is not `Expression.validate` per se, but calls to `hs_free_compile_error`, which also occur in `Database.handleErrors`.

**Comment by gliwka** [comment]
Yeah, thats true. 

I was able to reproduce it.

Basically it boils down to:
a) disabling autoread on the structs and freeing the structs using hs_free_compile_error
- or -
b) not freeing the struct and leaving it to the garbage collector, since the JNA implementation should take care of it

**Comment by gliwka** [comment]
```c
hs_error_t HS_CDECL hs_free_compile_error(hs_compile_error_t *error) {
#if defined(FAT_RUNTIME)
    if (!check_ssse3()) {
        return HS_ARCH_ERROR;
    }
#endif
    freeCompileError(error);
    return HS_SUCCESS;
}
```

That's the implementation of the hs_free_compile_error. In case it's a "fat library" (supporting multiple instruction sets), it makes sure at least sse3 is available. After that it calls     freeCompileError(error), which is implemented as:
```c
void freeCompileError(hs_compile_error_t *error) {
    if (!error) {
        return;
    }
    if (error == &hs_enomem || error == &hs_einternal ||
        error == &hs_badalloc) {
        // These are not allocated.
        return;
    }

    hs_misc_free(error->message);
    hs_misc_free(error);
}
```

hs_misc_free then just calls the default deallocator for the platform (free) or in case a user provided hyperscan with their own deallocator by using hs_set_misc_allocator() it then uses that.

**Comment by gliwka** [comment]
Since the memory is not allocated by JNA, it seems like it won't free the memory on garbage collection, so "disabling autoread on the structs and freeing the structs using hs_free_compile_error
or" is the proper solution.

I'm testing it right now and will push it later.

**Comment by gliwka** [comment]
Fixed in v0.4.11

**Comment by eliaslevy** [comment]
👍
